File: Challenges/NoRepeatSubstring.cs
namespace Challenges;

// Given a string s, find the length of the longest substring
// without repeating characters.

public class NoRepeatSubstringLength
{
    public static int Soln(string input)
    {
        int longestSoFar = 0;

        for (int i = 0; i < input.Length; i += 1)
        {
            List<char> seenChars = [];
            int lenCounter = 0;

            for (int j = i; j < input.Length; j += 1)
            {
                char currentChar = input[j];

                if (seenChars.Contains(currentChar))
                {
                    longestSoFar = (lenCounter > longestSoFar) ? lenCounter : longestSoFar ;
                    break;
                }
                else
                {
                    seenChars.Add(currentChar);
                    lenCounter += 1;
                }
            }

            longestSoFar = (lenCounter > longestSoFar) ? lenCounter : longestSoFar ;
        }

        return longestSoFar;
    }

    public static int Soln2(string input)
    {
        if (input.Length <= 1) return input.Length;

        Dictionary<char, int> map = [];

        int leftIndex = 0;
        int rightIndex = 0;

        int longestStreak = 0;
        int currentStreak = 0;

        while (rightIndex < input.Length)
        {
            char character = input[rightIndex];

            if (map.TryGetValue(character, out int characterCount))
            {
                if (characterCount == 0)
                {
                    // this is duplicated below; I would probably try to not duplicate it
                    map[character] = 1;
                    rightIndex += 1;
                    currentStreak += 1;
                    longestStreak = (currentStreak > longestStreak) ? currentStreak : longestStreak;
                }
                else if (characterCount == 1)
                {
                    map[character] += 1;

                    while (map[character] == 2)
                    {
                        if (leftIndex >= input.Length) return longestStreak;

                        char leftChar = input[leftIndex];

                        map[leftChar] -= 1;
                        currentStreak -= 1;

                        leftIndex += 1;
                    }

                    rightIndex += 1;
                    currentStreak += 1;
                }
                else
                {
                    throw new ApplicationException();
                }
            }
            else
            {
                map[character] = 1;
                rightIndex += 1;
                currentStreak += 1;
                longestStreak = (currentStreak > longestStreak) ? currentStreak : longestStreak;
            }
        }

        return longestStreak;
    }
}

File: javascript/noRepeatSubstring.js
// Given a string s, find the length of the longest 
// substring without repeating characters.

function solution(s)
{
    if (s.length <= 1) return s.length;

    let leftIndex = 0;
    let rightIndex = 0;

    let longestStreak = 0;
    let currentStreak = 0;

    let map = new Map();

    while (rightIndex < s.length)
    {
        let rightChar = s[rightIndex];
        
        if (!map.has(rightChar))
        {
            map.set(rightChar, 1);
            rightIndex += 1;
            currentStreak += 1;
            longestStreak = (currentStreak > longestStreak) ? currentStreak : longestStreak;
        }
        else
        {
            let rightCharAmt = map.get(rightChar);

            if (rightCharAmt == 0)
            {
                map.set(rightChar, 1);
                rightIndex += 1;
                currentStreak += 1;
                longestStreak = (currentStreak > longestStreak) ? currentStreak : longestStreak;
            }
            else if (rightCharAmt == 1)
            {
                map.set(rightChar, 2);

                while (map.get(rightChar) == 2)
                {
                    if (leftIndex >= s.length) return longestStreak;

                    let leftChar = s[leftIndex];

                    let leftCount = map.get(leftChar);

                    map.set(leftChar, leftCount - 1);

                    leftIndex += 1;
                    currentStreak -= 1;
                }

                rightIndex += 1;
                currentStreak += 1;
                longestStreak = (currentStreak > longestStreak) ? currentStreak : longestStreak;
            }
            else
            {
                throw "something went wrong";
            }
        }
    }

    return longestStreak; 
    
}

module.exports = solution;